### PR TITLE
Update spec; test docs snippets

### PR DESCRIPTION
**Pull request by MichaelCurrie**
(To handle part 2 of #34 and also #46)

Added a unit test to validate that all snippets in the `.md` docs validate against the schema, and also are able to be loaded by the `WCONWorms` class (which has a few additional checks like that the `x` and `y` data arrays have the same length per data segment.)

This required some small changes to the snippets, and in a couple of instances to the schema, based on my judgement of what we should consider "correct".

The changes to the schema were all to the `"metadata"` object, and were:
- allow anything in `"settings"`, not just strings
- `"age"` is a number not a string now.
- `"temperature": 22` instead of `"temperature": {"value": 22, "units": "C"}`
- `"humidity": 45` instead of `"humidity": {"value": 45, "units": "%"}`
- `"who"` allows for arrays of strings, not just strings, so we can have `"who": ["J. Smith", "Suzie Q."]`, in addition to `"who": "J.Smith"`

### CODE REVIEW

**Comment by MichaelCurrie** [comment]
@Ichoran if you approve you can merge this and close #34 and #46.  Thanks!

File: WCON_format.md [hunk 1]
```diff
@@ -280,12 +283,10 @@ An example WCON file with a complete metadata section is given below.
            "media":"NGM",
            "sex":"hermaphrodite",
            "stage":"dauer",
-           "age":38.4,
+           "age":18.511,
```
Comment (Ichoran) [inline WCON_format.md]: Why did you change this?  38.4 is a reasonable age (in hours) to get a dauer larva due to overcrowding.  Do you know a protocol that will give you dauers in 18.5 hours?  (High temperature?)

File: WCON_format.md [hunk 2]
```diff
@@ -280,12 +283,10 @@ An example WCON file with a complete metadata section is given below.
...
            "strain":"CB4856",
-           "protocol":[
-               "dauer induction by method in J. Doe, 'Get Dauers', J. of Stuff, v1 p234",
-               "worm transferred to arena 1-2 minutes before recording started"
-           ],
+           "protocol":
+               "dauer induction by method in J. Doe, 'Get Dauers', J. of Stuff, v1 p234\nworm transferred to arena 1-2 minutes before recording started",
```
Comment (Ichoran) [inline WCON_format.md]: I wanted this not to contain weird formatting characters like `\n` to seem more accessible.  Also, the `protocol` field is stated to allow arrays.

File: WCON_format.md [hunk 3]
```diff
@@ -255,7 +258,7 @@ As an example, inside
       "@XJ":{ "foo": { "e": 2 }, "yes": "I think so"},
       "settings":{"q": 4, "r": 5}
     },
-    "data": { "id":1, "t":0, "x":1, "y":2, "e":3, "@XJ": {"e": 3}, "f":{"q": 4} }
+    "data": [{ "id":1, "t":0, "x":[1], "y":[2], "@XJ": {"e": 3, "f":{"q": 4}}}]
```
Comment (Ichoran) [inline WCON_format.md]: Getting rid of the ignored `"e"` outside the `"@XJ"` is a good idea, but adding unnecessary brackets around the numbers for `x` and `y` doesn't make this easier to read, nor does making `data` an array when it's not required to be one.  Also, I don't think the text description quite matches any more.

File: WCON_format.md [hunk 4]
```diff
@@ -295,7 +296,7 @@ An example WCON file with a complete metadata section is given below.
      },
     "units":{"t":"s", "x":"mm", "y":"mm", "temperature":"C", "humidity":"%", "size":"mm", "age":"h"},
     "data":[
-        { "id":1, "t":1.3, "x":-5.3, "y":6.4, "@suzq":[true, true, false, true] }
+        { "id":1, "t":1.3, "x":[-5.3], "y":[6.4], "@suzq":[true, true, false, true] }
```
Comment (Ichoran) [inline WCON_format.md]: It's harder to read these with the brackets.  And this is meant to be a readable example.  Are they necessary?

File: WCON_format.md [hunk 5]
```diff
@@ -355,8 +356,7 @@ Here is an example WCON file using origin and centroid points:
     "units":{"t":"s", "x":"mm", "y":"mm"},
     "data":[
  	      { "id":1, "t":1.3, "x":[7.2, 8.1], "y":[0.5, 0.3],
-          "ox":32.4, "oy":9.2, "cx":7.676, "cy": 0.384
-        }
+          "ox":32.4, "oy":9.2, "@SmithLab": {"cx":7.676, "cy": 0.384}}
```
Comment (Ichoran) [inline WCON_format.md]: Centroid is part of the core specification.  Why is this in `@SmithLab` here?

File: WCON_format.md [hunk 6]
```diff
@@ -379,7 +379,7 @@ In this example, we imagine a directory containing files labelled `filename_0.wc
     "files":{
          "this":"_2",
          "prev":["_1", "_0"],
-         "next":"_3"
+         "next":["_3"]
```
Comment (Ichoran) [inline WCON_format.md]: I think this wasn't required to be an array.  Again, it's less readable.

File: src/Python/tests/tests.py [hunk 7]
```diff
@@ -381,15 +434,14 @@ def test_metadata(self):
...
-                       "humidity":{ "value":40, "units":"%" },
+                       "temperature":23.8,
+                       "humidity":40.3,
                        "dish":{ "type":"petri", "size":35, "units":"mm" },
                        "food":"none",
                        "media":"agarose",
                        "sex":"hermaphrodite",
                        "stage":"adult",
-                       "age":"18:25:43.511",
+                       "age":18.511,
```
Comment (Ichoran) [inline src/Python/tests/tests.py]: This was a poorly chosen example originally (I think my fault!)--you can't get an adult in 18.5 hours!

File: tests/minimax.wcon [hunk 8]
```diff
@@ -27,7 +27,10 @@
 		"ox":"mm",
 		"speed":"mm/s",
 		"curvature":"1/mm",
-		"width":"mm"
+		"width":"mm",
+		"humidity":"%",
+		"temperature":"C",
+        "age":"h"
```
Comment (Ichoran) [inline tests/minimax.wcon:33]: Indentation doesn't seem to match here.  Formatting everything with spaces helps keep things looking consistent for everyone.

File: src/Python/wcon/wcon_schema.json [hunk 9]
```diff
@@ -41,23 +41,13 @@
 			"description": "Information about experimental conditions and software versions; critical for reproducibility across labs.",
 			"properties": {
 				"lab": {"type": "object"},
-				"who": {"type": "string"},
+				"who": {"oneOf": [
```
Comment (Ichoran) [inline src/Python/wcon/wcon_schema.json]: Why does this file exist?  Shouldn't everyone just use the root-directory wcon_schema.json?

**Comment by Ichoran** [comment]
Mostly good, but I have a variety of fiddly comments.

File: WCON_format.md (hunk 1 above)
Comment (MichaelCurrie) [inline WCON_format.md]: Not being an actual researcher, you've caught me out on using unrealistic data.  Great point, I'll revise it back to your number.  I only changed it at all because I first changed it back to a string and then I just made up a number rather than reverting.

File: WCON_format.md (hunk 2 above)
Comment (MichaelCurrie) [inline WCON_format.md]: Oh I see - you are using arrays as a way of being able to do multiline strings in JSON.  OK, I will revise the schema to allow arrays in the `protocol` field and revert this example.

**Comment by MichaelCurrie** [comment]
On brackets: 

On brackets around data: certainly I could create a special case allowing for "data" to be either an array or an object, and so the object is just the same as a singleton array.  But I think it makes for fewer special cases to say that "data" is always an array, even if you only have a single entry.  If you feel strongly that the readability you get from dropping those brackets trumps schema simplicity, I can add "data" without brackets as an option to the schema.

As for the brackets on the "x" and "y" entries, this is again case of readability vs. schema simplicity.  Here "x" and "y" could have any finite number of entries, and if they had more than one entry, they would need to have those entries in an array.

I have ceded the readability vs. schema simplicity issue in the case of `"t"`: in that case,

`"t":1.3, "x":[1,2,3], "y":[3.2,3.5,3.6]`

in the single-value case, but the multivalue case requires arrays:

`"t":[1.1, 1.3], "x":[[1,2,3],[1,2,3]], "y":[[3.2,3.5,3.6],[3.2,3.5,3.6]]`

File: WCON_format.md (hunk 4 above)
Comment (MichaelCurrie) [inline WCON_format.md]: I have commented on the brackets issue in a comment at the bottom.

File: WCON_format.md (hunk 3 above)
Comment (MichaelCurrie) [inline WCON_format.md]: My schema explicitly disallows any entries other than `"id"`, `"t"`, `"x"`, `"y"`, and anything starting with `"@"`.  So `"e"` is not just ignored, it's not allowed.  I think that's better because we already allow custom WCON entries inside the allowed `"@"` fields.

I have commented on the brackets issue in a comment at the bottom.

File: WCON_format.md (hunk 6 above)
Comment (MichaelCurrie) [inline WCON_format.md]: I have commented on the brackets issue in a comment at the bottom.  In particular in this case, I've required arrays around previous and next, which I think in a way improves readability since it shows clearly that these entries accept multiple values, whereas "this" only has one value.

Certainly reasonable people can disagree on this point and if you feel it's less readable I can create an extra case in the schema to allow for single-valued "next" entries not to be wrapped in an array.

File: src/Python/tests/tests.py (hunk 7 above)
Comment (MichaelCurrie) [inline src/Python/tests/tests.py]: Aha, this must be the origin of my `18.511` figure from above as well.  I'll change to something around 33 hours instead.

File: tests/minimax.wcon (hunk 8 above)
Comment (MichaelCurrie) [inline tests/minimax.wcon:33]: Agree - I think this was a case of my editing this file with Notepad++ instead of my usual Spyder or vim.  Thanks for catching this; I'll fix it.

File: src/Python/wcon/wcon_schema.json (hunk 9 above)
Comment (MichaelCurrie) [inline src/Python/wcon/wcon_schema.json]: Very good question.  The reason is that the PyPI packager wraps up everything in the `src/Python/wcon/` folder and zips it up and that becomes what's distributed when you install wcon via `pip install wcon`.  Therefore since the schema must be included in the package, it must be in the `src/Python/wcon/` folder.

Of course, the schema is not Python-specific and so I wanted it to be in the root directory, as it is.  But that leaves us in the somewhat awkward position of having two identical `wcon_schema.json` files in two places in the repo.

I put a unit test run by Travis-CI which validates that both files are identical, so we'll never have a difference.

File: WCON_format.md (hunk 5 above)
Comment (MichaelCurrie) [inline WCON_format.md]: You're right - it looks like I somehow missed implementing centroid.  I only implemented origin ("ox", "oy").  I'll add centroid now.

File: tests/minimax.wcon (hunk 8 above)
Comment (MichaelCurrie) [inline tests/minimax.wcon:33]: I turns out indentation was inconsistent in `minimax.wcon`, the schema, and in `WCON_format.md`.  I've fixed this so all indentation is 4 spaces (no tabs).

**Comment by Ichoran** [comment]
I was hoping that we could adhere to the overall principle that "if you only have one of it, you don't need to bother putting it in an array".  This makes it more work for readers but less work for writers, and I'm most worried about other labs getting working WCON writers so they can convert their custom formats into WCON.  Thus, although I agree that it's not a big deal to require brackets for `data`, mandating it breaks the principle I had hoped we could adhere to.

Also, regarding failing if `e` exists--we say in the spec that readers must ignore things they don't understand.  So a wcon file should be totally valid if it has `{"id":0, "fish":true, "t":0, "x":0, "y":0}`.  Putting things in `@` blocks is purely for gaining some support: readers will keep the blocks, and will convert units within those blocks.  So, play nice and you get help.  Do your own weird stuff, and at least the basics which are there are still understandable.  So it's okay not to have that in any examples, I guess, but the Python reader should adhere to our own advice and ignore tags it doesn't understand (or at least have an option to ignore them).

File: WCON_format.md [hunk 10]
```diff
@@ -377,9 +379,9 @@ In this example, we imagine a directory containing files labelled `filename_0.wc
 ```JSON
 {
     "files":{
-         "this":"_2",
-         "prev":["_1", "_0"],
-         "next":"_3"
+        "this":"_2",
+        "prev":["_1", "_0"],
+        "next":["_3"]
```
Comment (Ichoran) [inline WCON_format.md:384]: Regardless of what we decide for brackets overall, here I think it's important to allow them to be dropped since the minimum needed is simply the prev and next item.  Any additional items are just for error-checking.  It's the very next and very previous ones that let you follow the chain, so a reader should only need to say them and ignore the possibility of having a comprehensive list in every file.

File: discuss/Formats.md [hunk 11]
```diff
@@ -53,16 +53,23 @@ JSON is a fairly minimal format that lets you deliver sets of labeled data.  The
...
 {
   "metadata": {
-    "lab":"Kerr",
-    "person":"Kerr",
+    "lab":{"location":"Janelia Research Campus", "name":"Kerr Lab" },
+    "who":"Kerr",
     "strain":"N2",
-    "date":"2015-04-27T15:15:32",
-    "units":"pixels"
+    "timestamp":"2015-04-27T15:15:32"
```
Comment (Ichoran) [inline discuss/Formats.md]: I wasn't at Janelia at this time.  If we're going to pick something plausible, we could make it 2011.  Or we can change to a fictitious person.  (I don't want to make people at UC San Francisco or Calico feel that I'm not appreciating the opportunity I have had / have to work there.)

File: src/Python/tests/tests.py [hunk 12]
```diff
@@ -381,15 +463,14 @@ def test_metadata(self):
...
-                       "humidity":{ "value":40, "units":"%" },
+                       "temperature":23.8,
+                       "humidity":40.3,
                        "dish":{ "type":"petri", "size":35, "units":"mm" },
                        "food":"none",
                        "media":"agarose",
                        "sex":"hermaphrodite",
                        "stage":"adult",
-                       "age":"18:25:43.511",
+                       "age":33.4,
```
Comment (Ichoran) [inline src/Python/tests/tests.py]: Better, but this is still super-young for an adult.  How about we double it?  66.8 or something should be comfortably into adulthood at 23.8C. Sorry to be so pedantic, but I think it makes a better impression on worm biologists if the examples make biological sense.

Comment (MichaelCurrie) [inline src/Python/tests/tests.py]: Sounds good, I agree we should have examples that make sense.

File: discuss/Formats.md (hunk 11 above)
Comment (MichaelCurrie) [inline discuss/Formats.md]: I've changed it to a fictitious person if that works.  ([Janet Wallace](http://memory-alpha.wikia.com/wiki/Janet_Wallace))

File: WCON_format.md (hunk 10 above)
Comment (MichaelCurrie) [inline WCON_format.md:384]: That makes sense.

**Comment by MichaelCurrie** [comment]
OK I will make changes to the schema and code to allow for the principle: ""if you only have one of it, you don't need to bother putting it in an array".

Also, I will remove the restriction in the schema on other items not conforming to the "@" recommendation.

File: discuss/Formats.md [hunk 13]
```diff
@@ -53,16 +53,23 @@ JSON is a fairly minimal format that lets you deliver sets of labeled data.  The
 ```JSON
 {
   "metadata": {
-    "lab":"Kerr",
-    "person":"Kerr",
+    "lab":{"location":"USS Enterprise", "name":"Biochemistry Lab" },
+    "who":"Janet Wallace",
```
Comment (Ichoran) [inline discuss/Formats.md:57]: This is great :)

**Comment by Ichoran** [comment]
Looks like tests are still failing, and I am not sure that the Python implementation adheres to the "centroid is used as offset if you do not specify an offset" rule.  But otherwise it looks good to me.

**Comment by MichaelCurrie** [comment]
The problem with the  "centroid is used as offset if you do not specify an offset" rule is that I drop the offset column upon loading, by adding the offset to the x and y coordinates.  But the centroid is data as well, so if there is no offset, the centroid is considered the offset but I cannot actually use it to offset the data and drop the centroid, since we want to keep information about the centroid.

If I don't drop the centroid but still use it to offset the data, repeated save/load cycles will cause the x and y coordinates to shift by the centroid amount every time it's loaded.

So instead, in the case of an offset being present, after adding the offset, I then subtract centroid from the x and y coordinates, so later code can always assume that if the centroid is present, the x and y coordinates are relative to the centroid.  This is the only way I can see to maintain idempotent save/load cycles.

**Comment by Ichoran** [comment]
You can also keep track of whether the offset was there to get idempotent save/loads, at the cost of having state that has to do with history instead of worm data.  But the scheme you described also sounds fine (i.e. if centroid is present, body positions are always relative to centroid).  The Python API is not really covered by the WCON specification.  Do you think centroid-relative X and Y is a sensible thing for users?

In the Scala code, that's what I use.  (And actually I compute a centroid if none is present, and the spine is relative to that.)

**Comment by MichaelCurrie** [comment]
Yes I think centroid-relative X and Y is fine, and even more human-readable.  If the process is automated, of course, the exact format matters less as long as it's standardised, so either way we are covered here.

I think my last commit should cover everything, so assuming the PR passes the CI tests, unless you see anything else feel free to merge.  Thanks for your detailed feedback on this one!
